[2201.7.0] - Add bal persist diagram to the overview page (#7147)
ballerina-platform/ballerina-dev-website: swan-lake/learn-the-platform/ballerina-persist/bal-persist-overview.md

keywords: ballerina, programming language, ballerina packages, persist, data model, cli tool, client api
permalink: /learn/ballerina-persist/persist-overview/
active: persist_overview

> intro: The `bal persist` feature allows you to store data in different data stores and retrieve them when needed. A data store can be a database, or an in-memory cache. The `bal persist` feature currently supports in-memory tables, MySQL and MSSQL databases, and Google Sheets as data stores. The important point is you can use the same syntax to access data in all these data stores. Therefore, you don't need to learn different syntaxes to access data in different data stores.

Reviewer: ```suggestion
intro: The `bal persist` feature allows you to store data in different data stores and retrieve them when needed. A data store can be a database or an in-memory cache. The `bal persist` feature currently supports in-memory tables, MySQL, MSSQL databases, and Google Sheets as data stores. As you can use the same syntax to access data in all these data stores, you do not need to learn different syntaxes to access data in different data stores.
```

> intro: The `bal persist` feature allows you to store data in different data stores and retrieve them when needed. A data store can be a database, an in-memory cache, or a file system. The `bal persist` feature currently supports in-memory tables, MySQL and MSSQL databases, and Google Sheets as data stores. The important point is you can use the same syntax to access data in all these data stores. Therefore, you don't need to learn different syntaxes to access data in different data stores.

Reviewer 2: We don't support file systems for persist. Should we remove that bit?

Author: Fixed in [c6eed56](https://github.com/ballerina-platform/ballerina-dev-website/pull/7147/commits/c6eed56baf0ced5943b8a546ee6055d38484f264)

redirect_from:
- /learn/ballerina-persist/persist-overview/
---

> This feature has three main components: Data Model, CLI tool, and Type-safe client API. 

Reviewer: ```suggestion
This feature has three main components: the data model, CLI tool, and type-safe client API. 
```

> * The data model definition is used to define the data model. 

Reviewer: ```suggestion
* Data model: the data model definition is used to define the data model. 
```
Shall we combine the above sentence with these 3 bullet points?

Author: Fixed in [d8d5667](https://github.com/ballerina-platform/ballerina-dev-website/pull/7147/commits/d8d5667f5300667aac35aaa9bfb7a0194be20ad4)

* The CLI tool is used to generate the client API for the data model. 
* The type-safe client API is used to access the data in the data store.

The following diagram illustrates how the `bal persist` feature works.

> ![bal persist diagram](/learn/images/bal-persist-diagram.png "bal persist diagram")

Reviewer 2: The image link seems to be broken?

```suggestion
![bal persist diagram](/public/learn/images/bal-persist-diagram.png "bal persist diagram")
```

Author: @praneesha do we need `public` part on the page link

Reviewer: No, we don't.

redirect_from:
- /learn/ballerina-persist/persist-overview/
---

> This feature has three main components: Data Model, CLI tool, and Type-safe client API. The data model is used to define the data model. The CLI tool is used to generate the client API for the data model. The type-safe client API is used to access the data in the data store.

Reviewer: ```suggestion
This feature has three main components: Data Model, CLI tool, and Type-safe client API. The data model is used to define the data model. The CLI tool is used to generate the client API for the data model. The type-safe client API is used to access the data in the data store.
```
Let's add 3 bullet points for the components.

The following diagram illustrates how the `bal persist` feature works.

![bal persist diagram](/learn/images/bal-persist-diagram.png "bal persist diagram")

> The data model is the source of truth for the data model in your application. Even though the data model is defined in Ballerina, it is not a Ballerina source file. It is a file that contains the data model definition. The data model is used to generate the client API which contains Ballerina client class and Ballerina types.

Reviewer: ```suggestion
The data model definition shown in the above diagram is the source of truth for the data model of your application. The data model definition is used only to generate the client API, which contains the Ballerina client class and types. It cannot be used directly in your applications. It is used to generate the client API.
```

> The Ballerina VSCode plugin supports to validation of the data model, provides code actions to create the data model, and visualizes the data model in an ER diagram.

Reviewer: ```suggestion
The Ballerina VS Code extension facilitates the validation, development (via code actions), and visualization (in an Entity Relationship diagram) of the data model definition when creating it.
```

> The client API is generated using the `bal persist` CLI tool. The client API is generated as a Ballerina source file. Therefore, you can easily import the client API into your Ballerina source files and use it in your business logic to access the data in the data store.

Reviewer: ```suggestion
The created data model definition will be used to generate the client API via the `bal persist` CLI tool as a Ballerina source file. You can import this generated client API into your Ballerina source files and use it in your business logic to access the data in the data store.
```

## CLI tool

> The `bal persist` CLI is used to generate the client API for the data model. Based on the data store, additional configurable files and setup scripts are generated. For example, if you are using relational database as the data store, the `persist_db_config.bal` file and the `script.sql` script are generated. The `persist_db_config.bal` file is used to configure the relational database connection. The `script.sql` file is used to create the table in the relational database. The client API is generated in the `generated` directory in the project root directory.

Reviewer 2: ```suggestion
The `bal persist` CLI is used to generate the client API for the data model. Based on the data store, additional configurable files and setup scripts are generated. For example, if you are using a relational database as the data store, the `persist_db_config.bal` file and the `script.sql` script are generated. The `persist_db_config.bal` file is used to configure the relational database connection. The `script.sql` file is used to create the tables in the relational database. The client API is generated in the `generated` directory in the project root directory.
```

> The `bal persist` CLI is used to generate the client API for the data model. Based on the data store, additional configurable files and setup scripts are generated. For example, if you are using a relational database as the data store, the `persist_db_config.bal` file and the `script.sql` script are generated. The `persist_db_config.bal` file is used to configure the relational database connection. The `script.sql` file is used to create the tables in the relational database. The client API is generated in the `generated` directory in the project root directory.

Reviewer: ```suggestion
The `bal persist` CLI is used to generate the client API for the data model. Based on the data store, additional configurable files and setup scripts are generated. For example, if you are using a relational database as the data store, the `persist_db_config.bal` file, and the `script.sql` script are generated. The `persist_db_config.bal` file is used to configure the relational database connection. The `script.sql` file is used to create the tables in the relational database. The client API is generated in the `generated` directory in the project root directory.
```
